src/net_processing.cpp

```diff
@@ -1370,6 +1375,17 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
             connman->PushMessage(pfrom, CNetMsgMaker(nSendVersion).Make("alert", finalAlert));
         }
 
+        // Ask this peer for the block headers of all of the blocks we know are invalid
+        // If we receive headers messages for these invalid blocks, we will ban them.
+        std::vector<uint256> invalid_block_hashes;
+        for (const std::pair<const uint256, CBlockIndex*>& item : mapBlockIndex)
```

Comment:
Should probably wait until VERACK for this.

Also, walking mapBlockIndex seems a little heavy here, no?

Reply:
Besides waling `mapBlockIndex`, what do you suggest? I was considering making something to just store the hashes of the invalid blocks instead.

Reply:
Yea, I'd say the fork-points need to be cached. Unsure if it'd be best to just do it here and be done with it, or maintain something in validation.cpp that (for ex) getchaintips could also use.

Reply:
Well it would have to be something that is loaded/generated on startup too.